### PR TITLE
[sudoers]: Align read-only commands with NetAAA

### PR DESCRIPTION
**Pull request by baijupn**
## Why I did it

SONiC's local sudoers policy must permit the diagnostic commands already authorized for NetAAA read-only accounts. Without matching local rules, those commands pass NetAAA authorization but fail at the host sudo layer.

## Work item tracking

Microsoft ADO (number only): 36758854

## How I did it

Added constrained `READ_ONLY_CMDS` entries for:

- `systemctl status`
- `TSC`
- `dmidecode -s system-product-name`
- `dmesg -D`
- `chage -l <user>`
- `docker exec swss md5sum /usr/bin/arp_update`

The `chage` rule restricts the username argument with a sudoers regex. `TSC` accepts no arguments, and the other commands are limited to their approved operations. The previously proposed `smartctl` command is not included. The `sonic_installer` alias and unrestricted `rvtysh` commands remain excluded.

## How to verify it

```bash
python3 -m unittest -v files.image_config.sudoers.tests.test_sudoers
bash -n files/image_config/sudoers/tests/test_sudoers_authorization.sh
bash files/image_config/sudoers/tests/test_sudoers_authorization.sh
SUDOERS_TEST_IMAGE=debian:bookworm-slim \
    bash files/image_config/sudoers/tests/test_sudoers_authorization.sh
make SONIC_BUILD_JOBS=4 target/sonic-vs.img.gz
```

The authorization suite verifies every approved command and representative denied variants on both Trixie and Bookworm. A full VS image build passed. The compressed QCOW2 image passed `qemu-img check`, and inspection of the installed `fs.squashfs` confirmed all command paths and sudoers rules. It also confirmed that the `smartctl` and `sonic_installer` additions are absent.

## Which release branch to backport

None requested.

## Tested branch

`master` plus commit `2d35cc42b1e5a9c4a6eedaf3bb046bd329813f7a`.

VS image SHA-256:

```text
1e25d47bb9393fbddc1b46dc7a1ba0134815b73f506c9a07320ed344c12cabb7
```

## Description for the changelog

Align constrained SONiC read-only sudo permissions with approved NetAAA diagnostic commands.

### CODE REVIEW

File: files/image_config/sudoers/tests/test_sudoers_authorization.sh [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+assert_denied /usr/bin/systemctl restart swss.service
+assert_denied /usr/bin/TSC no-stats
+assert_denied /usr/sbin/dmidecode -s system-serial-number
+assert_denied /usr/bin/dmesg -C
+assert_denied /usr/bin/chage -M 30 readonly
+assert_denied /usr/bin/chage -l readonly extra
+assert_denied /usr/bin/docker exec swss md5sum /usr/bin/other
+assert_denied /usr/bin/docker exec swss md5sum /usr/bin/arp_update extra
+assert_denied /usr/local/bin/sonic_installer list
+assert_denied /usr/bin/rvtysh -c configure
```
[the remark this answers is not in the document]
Comment (baijupn) [reply files/image_config/sudoers/tests/test_sudoers_authorization.sh]: This command is out of scope for this fix

File: files/image_config/sudoers/sudoers [hunk 2]
```diff
@@ -42,7 +48,8 @@ Cmnd_Alias      READ_ONLY_CMDS = /bin/cat /var/log/syslog, /bin/cat /var/log/sys
                                  /usr/local/bin/pcieutil *, \
                                  /usr/local/bin/psuutil *, \
                                  /usr/local/bin/sonic-installer list, \
-                                 /usr/local/bin/sfputil show *
+                                 /usr/local/bin/sfputil show *, \
```
[the remark this answers is not in the document]
Comment (baijupn) [reply files/image_config/sudoers/sudoers:51]: This command is not in scope for this change

File: files/image_config/sudoers/tests/test_sudoers.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    / "sonic_debian_extension.j2"
+)
+SMARTCTL_COMMAND = "/usr/sbin/smartctl -a /dev/sda"
+
+
+def read_command_alias(sudoers_text, alias_name):
+    lines = iter(sudoers_text.splitlines())
+    alias_prefix = f"Cmnd_Alias      {alias_name} ="
+
+    for line in lines:
```
[the remark this answers is not in the document]
Comment (baijupn) [reply files/image_config/sudoers/tests/test_sudoers.py]: This is standard formatting and not manually changing

File: files/image_config/sudoers/tests/test_sudoers.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    "/usr/bin/dmesg -D",
+    "/usr/bin/docker exec swss md5sum /usr/bin/arp_update",
+    "/usr/bin/systemctl status",
+    "/usr/bin/systemctl status *",
+    "/usr/sbin/dmidecode -s system-product-name",
+)
+EXCLUDED_COMMANDS = (
+    "/usr/local/bin/sonic_installer list",
+    "/usr/sbin/smartctl -a /dev/sda",
+)
```
[the remark this answers is not in the document]
Comment (baijupn) [reply files/image_config/sudoers/tests/test_sudoers.py]: removed exclusion tests

File: files/image_config/sudoers/tests/test_sudoers_authorization.sh [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    su -s /bin/sh readonly -c \
+        'sudo -n /usr/sbin/smartctl -a /dev/sda; id -u'
+)"
+test "$semicolon_output" = "$(printf '0|-a /dev/sda\n%s' "$readonly_uid")"
+
+and_output="$(
+    su -s /bin/sh readonly -c \
+        'sudo -n /usr/sbin/smartctl -a /dev/sda && id -u'
+)"
+test "$and_output" = "$(printf '0|-a /dev/sda\n%s' "$readonly_uid")"
```
[the remark this answers is not in the document]
Comment (baijupn) [reply files/image_config/sudoers/tests/test_sudoers_authorization.sh]: smartctl command is not in the scope for this and removed

File: files/image_config/sudoers/tests/test_sudoers_authorization.sh [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+EOF
```
Comment (qiluo-msft) [inline files/image_config/sudoers/tests/test_sudoers_authorization.sh]: Neither `test_sudoers.py` nor `test_sudoers_authorization.sh` is wired into any CI job — they live in the repo but nothing invokes them automatically. CodeQL, Semgrep, and the Azure build pipelines all pass without running them, so a future sudoers edit won't trigger a regression failure.

Two ways to close the gap:

1. **Hook `test_sudoers.py` into the existing CI** — the repo already has a `pytest.ini` and runs Python unit tests in CI for other components. Adding this to the test discovery path (or a lightweight GitHub Actions step) would give automatic coverage at zero infra cost.

2. **Add a case to `tests/tacacs/test_ro_user.py` in sonic-mgmt** — `test_ro_user_allowed_command` already has a `commands` dict that exercises each `READ_ONLY_CMDS` entry against a live DUT via TACACS. Adding `"smartctl": ["sudo smartctl -a /dev/sda"]` there (with the existing `does_command_exist` guard so it skips on NVMe/eMMC devices) would cover real enforcement end-to-end, consistent with how `pcieutil`, `lldpctl`, etc. are tested.

File: files/image_config/sudoers/sudoers [hunk 7]
```diff
@@ -22,13 +22,19 @@ Defaults        lecture_file = /etc/sudoers.lecture
 Cmnd_Alias      READ_ONLY_CMDS = /bin/cat /var/log/syslog, /bin/cat /var/log/syslog.1 /var/log/syslog, /bin/cat /var/log/syslog.1, \
                                  /sbin/ip netns identify [0-9]*, \
                                  /sbin/brctl show, \
+                                 /usr/bin/TSC "", \
+                                 /usr/bin/chage ^-l [A-Za-z0-9_.-]+$, \
+                                 /usr/bin/dmesg -D, \
                                  /usr/bin/docker exec snmp cat /etc/snmp/snmpd.conf, \
```
[the remark this answers is not in the document]
Comment (baijupn) [reply files/image_config/sudoers/sudoers:28]: this is evaluated and required to be supported

File: files/image_config/sudoers/tests/test_sudoers_authorization.sh (hunk 6 above)
Comment (baijupn) [reply files/image_config/sudoers/tests/test_sudoers_authorization.sh]: After discussing with Qi Luo, retaining only the tests in sonic-mgmt.

**Review by qiluo-msft — APPROVED** [review]

File: files/image_config/sudoers/sudoers [hunk 8]
```diff
@@ -22,13 +22,19 @@ Defaults        lecture_file = /etc/sudoers.lecture
 Cmnd_Alias      READ_ONLY_CMDS = /bin/cat /var/log/syslog, /bin/cat /var/log/syslog.1 /var/log/syslog, /bin/cat /var/log/syslog.1, \
                                  /sbin/ip netns identify [0-9]*, \
                                  /sbin/brctl show, \
+                                 /usr/bin/TSC "", \
```
[the remark this answers is not in the document]
Comment (baijupn) [reply files/image_config/sudoers/sudoers:25]: this needs parity in string match as policies expected does not require any arguments to be supported, this form is proper
